File: be/src/runtime/plugin/cloud_plugin_config_provider.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "runtime/plugin/cloud_plugin_config_provider.h"

#include <fmt/format.h>

#include "cloud/cloud_meta_mgr.h"
#include "cloud/cloud_storage_engine.h"
#include "cloud/config.h"
#include "runtime/exec_env.h"
#include "util/s3_util.h"

namespace doris {

Status CloudPluginConfigProvider::get_cloud_s3_config(
        std::unique_ptr<S3PluginDownloader::S3Config>* s3_config) {
    S3PluginDownloader::S3Config config("", "", "", "", "", "");
    Status status = _get_storage_vault_info(&config);
    RETURN_IF_ERROR(status);

    if (config.bucket.empty() || config.access_key.empty() || config.secret_key.empty()) {
        return Status::InvalidArgument(
                "Incomplete S3 configuration: bucket={}, access_key={}, secret_key={}",
                config.bucket, config.access_key.empty() ? "empty" : "***",
                config.secret_key.empty() ? "empty" : "***");
    }

    *s3_config = std::make_unique<S3PluginDownloader::S3Config>(
            config.endpoint, config.region, config.bucket, config.prefix, config.access_key,
            config.secret_key);
    return Status::OK();
}

Status CloudPluginConfigProvider::_get_storage_vault_info(S3PluginDownloader::S3Config* s3_config) {
    BaseStorageEngine& base_engine = ExecEnv::GetInstance()->storage_engine();
    CloudStorageEngine* cloud_engine = dynamic_cast<CloudStorageEngine*>(&base_engine);
    if (!cloud_engine) {
        return Status::NotFound("CloudStorageEngine not found, not in cloud mode");
    }

    cloud::CloudMetaMgr& meta_mgr = cloud_engine->meta_mgr();

    cloud::StorageVaultInfos vault_infos;
    bool enable_storage_vault = false;
    RETURN_IF_ERROR(meta_mgr.get_storage_vault_info(&vault_infos, &enable_storage_vault));

    if (vault_infos.empty()) {
        return Status::NotFound("No storage vault info available");
    }

    // Find first S3-compatible storage vault for plugin download
    for (const auto& [vault_id, vault_conf, path_format] : vault_infos) {
        if (const S3Conf* s3_conf = std::get_if<S3Conf>(&vault_conf)) {
            s3_config->endpoint = s3_conf->client_conf.endpoint;
            s3_config->region = s3_conf->client_conf.region;
            s3_config->bucket = s3_conf->bucket;
            s3_config->prefix = s3_conf->prefix;
            s3_config->access_key = s3_conf->client_conf.ak;
            s3_config->secret_key = s3_conf->client_conf.sk;

            LOG(INFO) << "Using storage vault for plugin download: " << vault_id
                      << ", prefix: " << s3_conf->prefix;
            return Status::OK();
        }
    }
    return Status::NotSupported("No S3-compatible storage found for plugin download");
}

} // namespace doris

File: fe/fe-core/src/main/java/org/apache/doris/common/plugin/CloudPluginConfigProvider.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.common.plugin;

import org.apache.doris.cloud.proto.Cloud;
import org.apache.doris.cloud.rpc.MetaServiceProxy;
import org.apache.doris.rpc.RpcException;

import com.google.common.base.Strings;

/**
 * CloudPluginConfigProvider retrieves S3 authentication info and plugin base paths
 * from Doris cloud mode environment.
 * <p>
 * Responsibilities:
 * - Configuration retrieval only, no download logic
 * - Converts complex cloud mode configuration to simple S3 config objects
 * - Provides unified configuration interface, hiding implementation details
 * <p>
 * Uses StorageVaultMgr to get default storage configuration, more stable and reliable.
 */
public class CloudPluginConfigProvider {

    /**
     * Get S3 configuration from cloud mode
     *
     * @return S3 config object
     * @throws RuntimeException if configuration retrieval fails
     */
    public static S3PluginDownloader.S3Config getCloudS3Config() {
        Cloud.ObjectStoreInfoPB objInfo = getStorageVaultInfo();

        if (Strings.isNullOrEmpty(objInfo.getBucket())
                || Strings.isNullOrEmpty(objInfo.getAk())
                || Strings.isNullOrEmpty(objInfo.getSk())) {
            throw new RuntimeException("Incomplete S3 configuration: bucket=" + objInfo.getBucket()
                    + ", ak=" + (Strings.isNullOrEmpty(objInfo.getAk()) ? "empty" : "***")
                    + ", sk=" + (Strings.isNullOrEmpty(objInfo.getSk()) ? "empty" : "***"));
        }

        return new S3PluginDownloader.S3Config(
                objInfo.getEndpoint(),
                objInfo.getRegion(),
                objInfo.getBucket(),
                objInfo.getPrefix(),
                objInfo.getAk(),
                objInfo.getSk()
        );
    }

    // ======================== Private Helper Methods ========================

    /**
     * Get storage configuration info for both SaaS and Vault modes
     * Uses the existing MetaServiceProxy.getObjStoreInfo() method directly
     */
    private static Cloud.ObjectStoreInfoPB getStorageVaultInfo() {
        try {
            Cloud.GetObjStoreInfoResponse response =
                    MetaServiceProxy.getInstance().getObjStoreInfo(Cloud.GetObjStoreInfoRequest.newBuilder().build());
            for (Cloud.StorageVaultPB vault : response.getStorageVaultList()) {
                if (vault.hasObjInfo()) {
                    return vault.getObjInfo();
                }
            }
            throw new RuntimeException("Failed to get obj store info");
        } catch (RpcException e) {
            throw new RuntimeException("Failed to get obj store info: " + e.getMessage());
        }
    }
}
